File: 240923/나무 타이쿤/tree-tycoon.py
from collections import deque


def move_supplements(d, p):
    global supplements
    temp = []
    dx, dy = move_dirs[d - 1]
    for x, y in supplements:
        nx = (x + dx * p) % n
        ny = (y + dy * p) % n
        temp.append((nx, ny))
    supplements = temp


def inject_supplements_and_growth():
    global tree_heights
    for x, y in supplements:
        tree_heights[x][y] += 1
    for x, y in supplements:
        for dx, dy in tree_dirs:
            nx = x + dx
            ny = y + dy
            if 0 <= nx < n and 0 <= ny < n and 1 <= tree_heights[nx][ny]:
                tree_heights[x][y] += 1


def cut_trees_and_add_supplements():
    global supplements, tree_heights
    temp = []
    for x in range(n):
        for y in range(n):
            if (x, y) not in supplements and 2 <= tree_heights[x][y]:
                tree_heights[x][y] -= 2
                temp.append((x, y))
    supplements = temp


move_dirs = ((0, 1), (-1, 1), (-1, 0), (-1, -1), (0, -1), (1, -1), (1, 0), (1, 1))  # → ↗ ↑ ↖ ← ↙ ↓ ↘
tree_dirs = ((-1, -1), (-1, 1), (1, 1), (1, -1))  # 나무 대각선
n, m = map(int, input().split())
supplements = [[0] * n for _ in range(n)]
tree_heights = [list(map(int, input().split())) for _ in range(n)]
move_rules = deque([tuple(map(int, input().split())) for _ in range(m)])
supplements = [(n - 1, 0), (n - 1, 1), (n - 2, 0), (n - 2, 1)]
for _ in range(m):
    d, p = move_rules.popleft()
    move_supplements(d, p)
    inject_supplements_and_growth()
    cut_trees_and_add_supplements()
result = 0
for th in tree_heights:
    result += sum(th)
print(result)

File: 240923/나무 타이쿤/tree-tycoon.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	
	static BufferedReader br;
	static StringTokenizer st;
	
	static final int[][] moveDirs = {{0, 1}, {-1, 1}, {-1, 0}, {-1, -1}, {0, -1}, {1, -1}, {1, 0}, {1, 1}};
	static final int[][] treeDirs = {{1, 1}, {1, -1}, {-1, 1}, {-1, -1}};
	
	static int n, m;
	static int[][] treeHeights;
	static List<Pos> supplements;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		treeHeights = new int[n][n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				treeHeights[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		supplements = new ArrayList<>();
		addInitialSupplements();
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());
			final int d = Integer.parseInt(st.nextToken()) - 1;
			final int p = Integer.parseInt(st.nextToken());
			moveSuppplements(d, p);
			injectSupplementsAndGrowth();
			cutTreesAndAddSupplements();
		}
		System.out.println(sumRemainigTreeHeights());
	}
	
	private static void addInitialSupplements() {
		supplements.add(new Pos(n - 1, 0));
		supplements.add(new Pos(n - 1, 1));
		supplements.add(new Pos(n - 2, 0));
		supplements.add(new Pos(n - 2, 1));
	}
	
	private static void moveSuppplements(final int d, final int p) {
		final List<Pos> temp = new ArrayList<>();
		for (final Pos supplement : supplements) {
			final int nx = (n + supplement.x + moveDirs[d][0] * p) % n;
			final int ny = (n + supplement.y + moveDirs[d][1] * p) % n;
			temp.add(new Pos(nx, ny));
		}
		supplements = temp;
	}

	private static void injectSupplementsAndGrowth() {
		for (final Pos supplenment : supplements) {
			treeHeights[supplenment.x][supplenment.y]++;
		}
		for (final Pos supplenment : supplements) {
			for (int d = 0; d < 4; d++) {
				final int nx = supplenment.x + treeDirs[d][0];
				final int ny = supplenment.y + treeDirs[d][1];
				if (0 <= nx && nx < n && 0 <= ny && ny < n && 1 <= treeHeights[nx][ny]) {
					treeHeights[supplenment.x][supplenment.y]++;
				}
			}
		}
	}

	private static void cutTreesAndAddSupplements() {
		final List<Pos> temp = new ArrayList<>();
		for (int x = 0; x < n; x++) {
			for (int y = 0; y < n; y++) {
				if (!supplements.contains(new Pos(x, y)) && 2 <= treeHeights[x][y]) {
					treeHeights[x][y] -= 2;
					temp.add(new Pos(x, y));
				}
			}
		}
		supplements = temp;
	}
	
	private static int sumRemainigTreeHeights() {
		int heights = 0;
		for (int x = 0; x < n; x++) {
			for (int y = 0; y < n; y++) {
				heights += treeHeights[x][y];
			}
		}
		return heights;
	}
	
	static class Pos {
		
		final int x;
		final int y;
		
		public Pos(final int x, final int y) {
			this.x = x;
			this.y = y;
		}
		
		@Override
		public boolean equals(Object object) {
			final Pos other = (Pos) object;
			return this.x == other.x && this.y == other.y;
		}
	}
}